netaddr.go

```diff
@@ -846,3 +871,422 @@ func (p *IPPrefix) UnmarshalText(text []byte) error {
+//
+// If either of r's bounds are invalid, in the wrong order, or if
+// they're of different address families, then Prefixes returns nil.
+func (r IPRange) Prefixes() []IPPrefix {
+	if !r.Valid() {
+		return nil
+	}
+	var makePrefix prefixMaker
+	if r.From.Is4() {
+		makePrefix = func(ip16 ip16, bits uint8) IPPrefix {
+			return IPPrefix{IPFrom16([16]byte(ip16)), bits - 12*8}
+		}
+	} else {
+		makePrefix = func(ip16 ip16, bits uint8) IPPrefix {
+			return IPPrefix{IPv6Raw([16]byte(ip16)), bits}
+		}
+	}
+	a16, b16 := ip16(r.From.As16()), ip16(r.To.As16())
+	return appendRangePrefixes(nil, makePrefix, a16, b16)
+}
+
+func appendRangePrefixes(dst []IPPrefix, makePrefix prefixMaker, a16, b16 ip16) []IPPrefix {
+	common := uint8(0)
+	for common < 128 && a16.bitSet(common) == b16.bitSet(common) {
+		common++
+	}
+	// See whether a16 and b16, after their common shared bits, end
+	// in all zero bits or all one bits, respectively.
+	aAllZero, bAllSet := true, true
+	for i := common; i < 128; i++ {
+		if a16.bitSet(i) {
+			aAllZero = false
+			break
+		}
+	}
+	for i := common; i < 128; i++ {
+		if !b16.bitSet(i) {
+			bAllSet = false
+			break
+		}
+	}
+	if aAllZero && bAllSet {
+		// a16 to b16 represents a whole range, like 10.50.0.0/16.
+		// (a16 being 10.50.0.0 and b16 being 10.50.255.255)
+		return append(dst, makePrefix(a16, common))
+	}
+	// Otherwise recursively do both halves.
+	dst = appendRangePrefixes(dst, makePrefix, a16, a16.lastWithBitZero(common+1))
+	dst = appendRangePrefixes(dst, makePrefix, b16.firstWithBitOne(common+1), b16)
+	return dst
+}
+
+func addOne(a []byte, i int) bool {
+	if v := a[i]; v < 0xff {
+		a[i]++
+		return true
+	}
+	if i == 0 {
+		return false
+	}
+	a[i] = 0
+	return addOne(a, i-1)
+}
+
+func subOne(a []byte, i int) bool {
+	if v := a[i]; v > 0 {
+		a[i]--
+		return true
+	}
+	if i == 0 {
+		return false
+	}
+	a[i] = 0xff
+	return subOne(a, i-1)
+}
+
+// ipFrom16Match returns an IP address from a with address family
+// matching ip.
+func ipFrom16Match(ip IP, a [16]byte) IP {
+	if ip.Is6() {
+		return IPv6Raw(a) // doesn't unwrap
+	}
+	return IPFrom16(a)
+}
+
+// Next returns the IP following ip.
+// If there is none, it returns the IP zero value.
+func (ip IP) Next() IP {
+	var ok bool
+	a := ip.As16()
+	if ip.Is4() {
+		ok = addOne(a[12:], 3)
+	} else {
+		ok = addOne(a[:], 15)
+	}
+	if ok {
+		return ipFrom16Match(ip, a)
+	}
+	return IP{}
+}
+
+// Prior returns the IP before ip.
+// If there is none, it returns the IP zero value.
+func (ip IP) Prior() IP {
+	var ok bool
+	a := ip.As16()
+	if ip.Is4() {
+		ok = subOne(a[12:], 3)
+	} else {
+		ok = subOne(a[:], 15)
+	}
+	if ok {
+		return ipFrom16Match(ip, a)
+	}
+	return IP{}
+}
+
+// IPRangeSet represents a set of IPRanges.
+//
+// The zero value is a valid value representing a set of no IP ranges.
+//
+// The Add and Remove methods add or remove ranges to the set.  Add
+// methods should be called first, as a remove operation does nothing
```

Comment:
This makes me a bit nervous. It seems too easy to misuse as an API. Maybe make the invariant stronger: Calling Remove after calling Add causes a panic. Or put another way: You must add all desired ranges, then remove any undesired ranges.